tiledb/sm/rest/rest_profile.cc

```diff
@@ -0,0 +1,300 @@
+    write_file(data, filepath_);
+  }
+}
+
+json RestProfile::to_json() {
+  json j;
+  for (const auto& param : param_values_) {
+    j[param.first] = param.second;
+  }
+  return j;
+}
+
+std::string RestProfile::dump() {
+  return json{{name_, to_json()}}.dump(2);
+}
+
+/* ****************************** */
+/*           PRIVATE API          */
+/* ****************************** */
+
+void RestProfile::load_from_json_file(const std::string& filename) {
+  if (filename.empty() ||
+      (filename != filepath_ && filename != old_filepath_)) {
+    throw RestProfileException(
+        "Cannot load from \'" + filename + "\'; invalid filename.");
+  }
+
+  if (!std::filesystem::exists(filename)) {
+    throw RestProfileException(
+        "Cannot load from \'" + filename + "\'; file does not exist.");
+  }
+
+  // Load the file into a json object.
+  json data = read_file(filename);
+
+  // Update any written-parameters from the loaded json object.
+  if (filename.c_str() == old_filepath_.c_str()) {
+    if (data.contains("api_key") &&
+        data["api_key"].contains("X-TILEDB-REST-API-KEY")) {
+      param_values_["rest.token"] = data["api_key"]["X-TILEDB-REST-API-KEY"];
+    }
+    if (data.contains("host")) {
+      param_values_["rest.server_address"] = data["host"];
+    }
+    if (data.contains("password")) {
+      param_values_["rest.password"] = data["password"];
+    }
+    if (data.contains("username")) {
+      param_values_["rest.username"] = data["username"];
+    }
+  } else {
+    json profile = data[name_];
+    if (!profile.is_null()) {
+      for (auto it = profile.begin(); it != profile.end(); ++it) {
+        param_values_[it.key()] = profile[it.key()];
+      }
+    }
+  }
+}
+
+json RestProfile::read_file(const std::string& filepath) {
+  // Temporarily append filepath with a random label to guarantee atomicity.
+  std::string temp_filepath = filepath;
+  if (std::filesystem::exists(filepath)) {
+    temp_filepath += "." + random_label();
+    if (std::rename(filepath.c_str(), temp_filepath.c_str()) != 0) {
+      throw RestProfileException(
+          "Failed to load file due to an internal error.");
+    }
+  }
+
+  // Read the file into a json object.
+  json data;
+  {
+    std::ifstream file(temp_filepath);
+    try {
+      file >> data;
+    } catch (...) {
+      throw RestProfileException(
+          "Error parsing file \'" + temp_filepath + "\'.");
+    }
+  }
+
+  // Remove the random label from the filepath, if applicable.
+  if (temp_filepath != filepath) {
+    if (std::rename(temp_filepath.c_str(), filepath.c_str()) != 0) {
+      throw RestProfileException(
+          "Failed to load file due to an internal error.");
+    }
+  }
+
+  return data;
+}
+
+void RestProfile::write_file(json data, const std::string& filepath) {
+  // Temporarily append filepath with a random label to guarantee atomicity.
+  std::string temp_filepath = filepath;
+  if (std::filesystem::exists(filepath)) {
```

Comment:
We should remove this existence check. The point isn't just for when the file exists, its that we're never actually writing once its renamed. With this check it leaves open the (admittedly tiny) chance that we're writing the first version when some other process/thread attempts to read from it.